src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,296 @@
+     * ahead of time. These will be later used by the inclusive scan operator.
+     **/
+    if (ret.is_cat) {
+      ret.computed_result.left_sum = split_input.parent_sum - ret.gpair;
+      ret.computed_result.right_sum = ret.gpair;
+    } else {
+      if (forward) {
+        ret.computed_result.left_sum = ret.gpair;
+        ret.computed_result.right_sum = split_input.parent_sum - ret.gpair;
+      } else {
+        ret.computed_result.left_sum = split_input.parent_sum - ret.gpair;
+        ret.computed_result.right_sum = ret.gpair;
+      }
     }
-    cub::CTA_SYNC();
+    ret.computed_result.parent_sum = split_input.parent_sum;
+    float parent_gain = evaluator.CalcGain(split_input.nidx, split_input.param,
+                                           GradStats{ret.computed_result.parent_sum});
+    float gain = evaluator.CalcSplitGain(split_input.param, split_input.nidx, ret.findex,
+                                         GradStats{ret.computed_result.left_sum},
+                                         GradStats{ret.computed_result.right_sum});
+    ret.computed_result.best_loss_chg = gain - parent_gain;
+    ret.computed_result.best_findex = ret.findex;
+    ret.computed_result.best_fvalue = ret.fvalue;
+    ret.computed_result.best_direction =
+        (forward ? DefaultDirection::kRightDir : DefaultDirection::kLeftDir);
   }
-}
-
-template <int BLOCK_THREADS, typename GradientSumT>
-__global__ void EvaluateSplitsKernel(
-    EvaluateSplitInputs<GradientSumT> left,
-    EvaluateSplitInputs<GradientSumT> right,
-    TreeEvaluator::SplitEvaluator<GPUTrainingParam> evaluator,
-    common::Span<DeviceSplitCandidate> out_candidates) {
-  // KeyValuePair here used as threadIdx.x -> gain_value
-  using ArgMaxT = cub::KeyValuePair<int, float>;
-  using BlockScanT =
-      cub::BlockScan<GradientSumT, BLOCK_THREADS, cub::BLOCK_SCAN_WARP_SCANS>;
-  using MaxReduceT = cub::BlockReduce<ArgMaxT, BLOCK_THREADS>;
 
-  using SumReduceT = cub::BlockReduce<GradientSumT, BLOCK_THREADS>;
-
-  union TempStorage {
-    typename BlockScanT::TempStorage scan;
-    typename MaxReduceT::TempStorage max_reduce;
-    typename SumReduceT::TempStorage sum_reduce;
-  };
+  return ret;
+}
 
-  // Aligned && shared storage for best_split
-  __shared__ cub::Uninitialized<DeviceSplitCandidate> uninitialized_split;
-  DeviceSplitCandidate& best_split = uninitialized_split.Alias();
-  __shared__ TempStorage temp_storage;
+template <typename GradientSumT>
+__noinline__ __device__ thrust::tuple<ScanElem<GradientSumT>, ScanElem<GradientSumT>>
+ScanValueOp<GradientSumT>::operator() (
+    thrust::tuple<EvaluateSplitsHistEntry, EvaluateSplitsHistEntry> entry_tup) {
+  const auto& fw = thrust::get<0>(entry_tup);
+  const auto& bw = thrust::get<1>(entry_tup);
+  ScanElem<GradientSumT> ret_fw, ret_bw;
+  ret_fw = MapEvaluateSplitsHistEntryToScanElem<true>(
+      fw,
+      (fw.indicator == ChildNodeIndicator::kLeftChild ? this->left : this->right));
```

Comment:
In the future, we might want to support multiple nodes.  Since the size of the histogram is fixed for all nodes I think a division can obtain the node index inside node set.

Reply:
Renamed the field to `node_idx`